### ISSUE TITLE
Resolving either ECDSA or RSA CertifiedKey

### ISSUE THREAD

**Issue by jeromegn** [ctz/rustls#297]
I've started integrating rustls in our TCP proxy and I've stumbled on this little issue when translating code from OpenSSL.

With OpenSSL, you can add certificates and it'll make an informed decision on which to choose. I tried implementing something similar within my implementation of `ResolvesServerCert`, but I have not been able to properly filter certificates.

We're generating both ECDSA and RSA certificates via Let's Encrypt for every domain we proxy. I can't tell if I should be returning ECDSA or RSA based on the information available in the `resolve` function arguments.

The `sigschemes` is almost always the same and includes both ECDSA and RSA schemes, whatever ciphersuite is requested. I tried both with curl and openssl s_client:

```
curl -ivk https://127.0.0.1:8443 --ciphers ECDHE-RSA-AES128-GCM-SHA256 --tls-max 1.2
```

```
openssl s_client -debug -connect 127.0.0.1:8443 -cipher 'RSA'
```

If I return an ECDSA `CertifiedKey` from my implementation's `resolve` function with these requests, I'll get a rustls error:

```
peer is incompatible: no ciphersuites in common
```

I tried digging through the code, but couldn't figure out if there was some sort of filtering going on with the signature schemes, other than based on the protocol.

Not all ciphersuites can be signed with either RSA and ECDSA (in fact, they're pretty rare).

Is there a way to make an informed decision to sign with either a ECDSA or RSA CertifiedKey that I haven't figured out?

I'd be happy with a list of ciphers in the `resolve` function signature. Or a filtered version of which signature schemes are actually supported (and maybe a way to test if the scheme works for a given CertifiedKey?)

**Comment by ctz** [comment]
Thanks for the report. Yes, this is a bug. The intention is that `sigschemes` comes from the client, but that needs to be reduced by the offered ciphersuites for TLS1.2 otherwise you might choose a key which is unusable with any offered suite. I will prepare a fix and some tests for this case.

**Comment by jeromegn** [comment]
Thanks, that's what I imagined was the intended behavior!

As I'm implementing more of this, I realized I needed to know which ALPN the client sent from the `resolve` function. This is to provide an answer to the tls-alpn-01 Let's Encrypt challenge. Essentially I need to be able to match "acme-tls/1" from the client's ALPN protocols.

I made a quick branch that exposes the raw `ClientHelloPayload` to the `resolve` method, but this is not great as it doesn't provide friendly functions and structs. It does work for our needs though.

I figure a proper implementation would create a facade struct for this payload. I'm not sure how we could modify the `resolve` function signature without a breaking change though.

I should open a different issue for this... :smile:
